### PR TITLE
bpo-31116: Add Z85 variant to base64

### PR DESCRIPTION
**Pull request by matan1008**
https://bugs.python.org/issue31116

* #75299

### CODE REVIEW

**Review by MaxwellDupre — CHANGES_REQUESTED** [review]
This needs some documentation a single short NEWS entry I don't think is enough. A short description plus maybe a link to reference.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
I apologize for the fact that this PR was not reviewed earlier. The idea itself looks reasonable to me, and the implementation looks clever, so I am going to approve this change.

Please add a variant of test_b85decode_errors for Z85. I suspect that there may be some issues here. Also please add a What's New entry and update `versionadded` directives.

**Comment by matan1008** [comment]
I have made the requested changes; please review again

File: Lib/test/test_base64.py [hunk 1]
```diff
@@ -776,6 +776,20 @@ def test_b85decode_errors(self):
...
+                  list(range(128, 256))
+        for c in illegal:
+            with self.assertRaises(ValueError, msg=bytes([c])):
+                base64.z85decode(b'0000' + bytes([c]))
+
+        self.assertRaises(ValueError, base64.z85decode, b'%')
+        self.assertRaises(ValueError, base64.z85decode, b'%N')
+        self.assertRaises(ValueError, base64.z85decode, b'%Ns')
+        self.assertRaises(ValueError, base64.z85decode, b'%NsC')
+        self.assertRaises(ValueError, base64.z85decode, b'%NsC1')
```
Comment (serhiy-storchaka) [inline Lib/test/test_base64.py]: The first four should be prefixes of `b'%nSc0'` which is a Z85 encoded `b'\xff\xff\xff\xff'`. The last one should be `b'%nSc1'`, it represents 0x100000000 which cannot be packed in 4 bytes.

Perhaps it needs a comment.

Comment (matan1008) [reply Lib/test/test_base64.py]: Fixed, Thanks!

File: Lib/test/test_base64.py [hunk 2]
```diff
@@ -707,14 +776,30 @@ def test_b85decode_errors(self):
...
 
+    def test_z85decode_errors(self):
+        illegal = list(range(33)) + \
+                  list(b'"\',;_`|\\~') + \
+                  list(range(128, 256))
+        for c in illegal:
+            with self.assertRaises(ValueError, msg=bytes([c])):
+                base64.z85decode(b'0000' + bytes([c]))
+
+        # b'\xff\xff\xff\xff' encodes to b'%nSc1', the following will overflow:
```
Comment (serhiy-storchaka) [inline Lib/test/test_base64.py]:
```suggestion
        # b'\xff\xff\xff\xff' encodes to b'%nSc0', the following will overflow:
```

File: Doc/whatsnew/3.13.rst [hunk 3]
```diff
@@ -224,6 +224,14 @@ asyncio
   the buffer size.
   (Contributed by Jamie Phan in :gh:`115199`.)
 
+base64
+---
+
+* Add :func:`base64.z85encode` and :func:`base64.z85decode` functions which allow encoding
+  and decoding z85 data.
+  See `Z85  specification <https://rfc.zeromq.org/spec/32/>`_ for more information.
+  (Contributed by Matan Perelman in :issue:`30598`.)
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.13.rst]: It should refer to the issue number, not the PR number.
```suggestion
  (Contributed by Matan Perelman in :issue:`75299`.)
```

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

For future, please do not use force push during review. It makes more difficult to see the incremental changes and requires to start reviewing from the start after every change.

File: Doc/whatsnew/3.13.rst [hunk 4]
```diff
@@ -224,6 +224,14 @@ asyncio
   the buffer size.
   (Contributed by Jamie Phan in :gh:`115199`.)
 
+base64
+---
+
+* Add :func:`base64.z85encode` and :func:`base64.z85decode` functions which allow encoding
+  and decoding z85 data.
+  See `Z85  specification <https://rfc.zeromq.org/spec/32/>`_ for more information.
+  (Contributed by Matan Perelman in :issue:`75299`.)
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.13.rst]:
```suggestion
  (Contributed by Matan Perelman in :gh:`75299`.)
```

**Comment by serhiy-storchaka** [comment]
Thank you for your contribution @matan1008.

**Comment by matan1008** [comment]
Thank you for accepting it!

**Pull request merged**